daemon/mgr/container_copy.go

```diff
@@ -0,0 +1,285 @@
+package mgr
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/alibaba/pouch/apis/types"
+	"github.com/alibaba/pouch/pkg/ioutils"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/chrootarchive"
+	"github.com/docker/docker/pkg/mount"
+	"github.com/go-openapi/strfmt"
+)
+
+// StatPath stats the dir info at the specified path in the container.
+func (mgr *ContainerManager) StatPath(ctx context.Context, name, path string) (stat *types.ContainerPathStat, err error) {
+	c, err := mgr.container(name)
+	if err != nil {
+		return nil, err
+	}
+	c.Lock()
+	defer c.Unlock()
+
+	running := c.IsRunningOrPaused()
+	err = mgr.Mount(ctx, c, false)
+	if err != nil {
+		return nil, err
+	}
+
+	defer mgr.Unmount(ctx, c, false, !running)
+
+	if !running {
+		err = mgr.attachVolumes(ctx, c)
+		if err != nil {
+			return nil, err
+		}
+		defer mgr.detachVolumes(ctx, c, false)
+	}
+
+	err = c.mountVolumes(!running)
+	if err != nil {
+		return nil, err
+	}
+	defer c.unmountVolumes(!running)
+
+	resolvedPath, absPath := c.getResolvedPath(path)
+	lstat, err := os.Lstat(resolvedPath)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.ContainerPathStat{
+		Name:  lstat.Name(),
+		Path:  absPath,
+		Size:  strconv.FormatInt(lstat.Size(), 10),
+		Mode:  uint32(lstat.Mode()),
+		Mtime: strfmt.DateTime(lstat.ModTime()),
+	}, nil
+}
+
+// ArchivePath return an archive and dir info at the specified path in the container.
+func (mgr *ContainerManager) ArchivePath(ctx context.Context, name, path string) (content io.ReadCloser, stat *types.ContainerPathStat, err error) {
+	c, err := mgr.container(name)
+	if err != nil {
+		return nil, nil, err
+	}
+	c.Lock()
+	defer func() {
+		if err != nil {
```

Comment:
recommend to use `err0`(named return value) instead of `err` here. `err0` can help us to distinguish with normal error